### ISSUE TITLE
MemberNameEqualsClassName factory function exemption doesn't work on generics

### ISSUE THREAD

**Issue by josephlbarnett** [detekt/detekt#3591]
## Expected Behavior
The `MemberNameEqualsClassName` rule is documented as `Factory functions that create an instance of the class are exempt from this rule.`  Would expect this exemption to apply even if the class in question has generics.

## Observed Behavior
A generic class with a generic factory function is reported as violating the rule.

## Steps to Reproduce
This code triggers the rule:
```kotlin
data class FieldInfo<T>(
    val alias: String,
    val extractor: (String) -> T
) {
    companion object {
        // this complains about MemberNameEqualsClassName
        fun <T> fieldInfo(
            alias: String,
            extractor: (String) -> T
        ): FieldInfo<T> {
            return FieldInfo(alias, extractor)
        }
    }
}
```
Removing the generic does not trigger the rule:
```kotlin
data class FieldInfo(
    val alias: String,
    val extractor: (String) -> String
) {
    companion object {
        // this does not complain about MemberNameEqualsClassName
        fun fieldInfo(
            alias: String,
            extractor: (String) -> String
        ): FieldInfo {
            return FieldInfo(alias, extractor)
        }
    }
}
```

## Context
The above example is a trivial reproduction, in our real world use case we have inline factory functions with reified type parameters that provide default `extractor` implementations. (eg `inline fun <reified T> fieldInfo(alias: String) = FieldInfo(alias, defaultImpl())` where `defaultImpl` is an inline function that returns different implementations based on T::class)

## Your Environment

* Version of detekt used:  1.16.0
* Version of Gradle used (if applicable): n/a, maven/antrun config here: https://github.com/trib3/leakycauldron/blob/main/parent-pom/pom.xml#L1469-L1505
* Gradle scan link (add `--scan` option when running the gradle task): n/a
* Operating System and version: ubuntu 21.04
* Link to your project (if it's a public repository): n/a, but public maven parent pom linked above

**Comment by josephlbarnett** [comment]
It appears that `typeReference.text == klass.name` (https://github.com/detekt/detekt/blob/main/detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/MemberNameEqualsClassName.kt#L104) is `false` because `typeReference.text` evaluates to `FieldInfo<T>` and `klass.name` evaluates to `FieldInfo`

**Comment by BraisGabin** [comment]
Thanks for the research! Do you have the time to provide a PR to fix it?
